cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java

```diff
@@ -169,6 +205,9 @@ public ExecutionSparkContext(SparkSpecification specification, Id.Program progra
     T dataset = (T) datasets.get(key);
     if (dataset == null) {
       dataset = instantiateDataset(name, datasetArgs);
+      if (dataset instanceof TransactionAware) {
+        ((TransactionAware) dataset).startTx(getTransaction());
+      }
```

Comment:
This was needed because if a dataset is instantiated in the middle of a spark job, it needs the transaction. Apparently we do not have existing test cases for Spark jobs with datasets.